src/appleseed/renderer/kernel/rendering/rendererservices.cpp

```diff
@@ -343,36 +367,57 @@ bool RendererServices::transform_points(
 
             // Convert from camera to world.
             for (int i = 0; i < npoints; ++i)
-                Pout[i] = Pin[i] * m;
+                transform(Pin[i], m, vectype, Pout[i]);
 
             from = g_world_ustr;
         }
         else return false;
 
         // Transform to NDC.
-        for (size_t i = 0; i < npoints; ++i)
+        if (vectype == OSL::TypeDesc::VECTOR)
+        {
+            Vector2d ndc_p(-1.0, -1.0);
+
+            if (sg)
+            {
+                m_camera->project_point(
+                    time,
+                    Vector3d(sg->P.x, sg->P.y, sg->P.z),
+                    ndc_p);
+            }
+
+            for (size_t i = 0; i < npoints; ++i)
+            {
+                OSL::Vec3 Q = sg->P + Pout[i];
+
+                Vector2d ndc_q(-1.0, -1.0);
+                m_camera->project_point(time, Vector3d(Q.x, Q.y, Q.z), ndc_q);
+                Pout[i] = OSL::Vec3(
+                    static_cast<float>(ndc_q[0] - ndc_p[0]),
+                    static_cast<float>(ndc_q[1] - ndc_p[1]),
+                    0.0f);
+            }
+        }
+        else // if (vectype == OSL::TypeDesc::POINT)
         {
-            Vector2d ndc;
-            if (m_camera->project_point(time, Vector3d(Pout[0].x, Pout[0].y, Pout[0].z), ndc))
+            for (size_t i = 0; i < npoints; ++i)
             {
+                Vector2d ndc(-1.0, -1.0);
+                m_camera->project_point(time, Vector3d(Pout[i].x, Pout[i].y, Pout[i].z), ndc);
                 Pout[i] = OSL::Vec3(
                     static_cast<float>(ndc[0]),
                     static_cast<float>(ndc[1]),
                     0.0f);
             }
-            else
-                Pout[i] = OSL::Vec3(-1.0f, -1.0f, 0.0f);
         }
 
         if (to == g_raster_ustr)
         {
             // Transform from NDC to raster.
             for (int i = 0; i < npoints; ++i)
```

Comment:
What's the type of `npoints`? I haven't checked but maybe `i` needs to be of type `size_t`.